DOCSP-37496 - FAQ (#35)
mongodb/docs-pymongo: source/faq.txt

Author: Do you think it makes sense to have "On this page" here, or does that clutter it up too much? My gut is to remove.

Reviewer: I like having it on there since it lets people skim the questions really quick to see if there's one on there that they want to read more about

.. meta::
   :keywords: errors, problems, help, troubleshoot

Is PyMongo Thread-Safe?
Is {+driver-short+} Thread-Safe?
-----------------------

PyMongo is thread-safe and provides built-in connection pooling
Yes. {+driver-short+} is thread-safe and provides built-in connection pooling
for threaded applications.

.. _pymongo-fork-safe:

Is PyMongo Fork-Safe?
Is {+driver-short+} Fork-Safe?
---------------------

PyMongo is not fork-safe. Take care when using instances of
``~pymongo.mongo_client.MongoClient`` with ``fork()``. Specifically,
don't copy instances of ``MongoClient`` from a parent process to
a child process. Instead, the parent process and each child process must
create their own instances of ``MongoClient``. Instances of ``MongoClient`` copied from
the parent process are likely to deadlock in the child process. {+driver-short+} will
attempt to issue a warning if there is a chance of this deadlock occurring.

.. _pymongo-fork-safe-details:

> ``MongoClient`` spawns multiple threads to run background tasks, such as monitoring

Author: moved to troubleshooting

To use multiprocessing with {+driver-short+}, write code similar to the following example:

.. code-block:: python

   # Each process creates its own instance of MongoClient.
   def func():
       db = pymongo.MongoClient().mydb
       # Do something with db.

   proc = multiprocessing.Process(target=func)
   proc.start()

.. important::
   
   Do not copy an instance of the ``MongoClient`` class from the parent process to a child
   process.

Can {+driver-short+} Load the Results of a Query as a Pandas DataFrame?
-----------------------------------------------------------------------

> You can use the `PyMongoArrow <https://mongo-arrow.readthedocs.io/en/pymongoarrow-0.1.1/>`__

Author: ah right

library to work with numerical or columnar data. PyMongoArrow lets you 
load MongoDB query result-sets as
`Pandas DataFrames <https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html>`__,
`NumPy ndarrays <https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html>`__, or
`Apache Arrow Tables <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`__.

How Does Connection Pooling Work in {+driver-short+}?

> --------------------------------------------

Author: Copied this section from the Node FAQ

``MongoClient`` supports multiple concurrent requests. For each process,
create a client and reuse it for all operations in a process. This
practice is more efficient than creating a client for each request.

The driver does not limit the number of requests that
can wait for sockets to become available, and it is the application's
responsibility to limit the size of its pool to bound queuing
during a load spike. Requests wait for the amount of time specified in
the ``waitQueueTimeoutMS`` option, which defaults to ``0`` (no limit).

A request that waits more than the length of time defined by
``waitQueueTimeoutMS`` for a socket raises a ``ConnectionFailure`` error. Use this
option if it is more important to bound the duration of operations
during a load spike than it is to complete every operation.

When ``MongoClient.close()`` is called by any request, the driver
closes all idle sockets and closes all sockets that are in
use as they are returned to the pool. Calling ``MongoClient.close()``
closes only inactive sockets, so you cannot interrupt or terminate
any ongoing operations by using this method. The driver closes these
sockets only when the process completes.

> For more information, see the :manual:`<administration/connection-pool-overview/>`

Reviewer: Broken link:

```suggestion
For more information, see the :manual:`</administration/connection-pool-overview/>`
```

  itself, but it wouldn't report the value back to {+driver-short+} for your application
  to use.
- Copying the document before adding the ``_id`` field would be
  prohibitively expensive for most high-write-volume applications.

If you don't want {+driver-short+} to add an ``_id`` to your documents, insert only
documents that already have an ``_id`` field, added by your application.

You might discover this behavior when calling
the ``Collection.insert_many()`` method with a list of references
to a single document. This raises a ``BulkWriteError``. Several
Python idioms can lead to this pitfall:

.. code-block:: python

   >>> doc = {}
   >>> collection.insert_many(doc for _ in range(10))
   Traceback (most recent call last):
   ...
   pymongo.errors.BulkWriteError: batch op errors occurred
   >>> doc
   {'_id': ObjectId('560f171cfba52279f0b0da0c')}

   >>> docs = [{}]
   >>> collection.insert_many(docs * 10)
   Traceback (most recent call last):
   ...
   pymongo.errors.BulkWriteError: batch op errors occurred
   >>> docs
   [{'_id': ObjectId('560f1933fba52279f0b0da0e')}]

Why Does My Query Work in the Shell But Not in {+driver-short+}?

> -------------------------------------------------------

Author: Moved to Troubleshooting

   >>> collection.find_one({'subdocument.a': 1.0,
   ...                      'subdocument.b': 1.0})
   {'_id': 1.0, 'subdocument': {'a': 1.0, 'b': 1.0}}

The query matches any subdocument with an "a" of 1.0 and a "b" of 1.0,
regardless of the order in which you specify them in Python, or the order in which they're
stored in BSON. This query also now matches subdocuments with additional
keys besides "a" and "b", whereas the previous query required an exact match.

The second solution is to use a ``~bson.son.SON`` object to specify the key order:

.. code-block:: python

   >>> query = {'subdocument': SON([('b', 1.0), ('a', 1.0)])}
   >>> collection.find_one(query)
   {'_id': 1.0, 'subdocument': {'a': 1.0, 'b': 1.0}}

The key order you use when you create a ``~bson.son.SON`` is preserved
when it is serialized to BSON and used as a query. Thus, you can create a
subdocument that exactly matches the subdocument in the collection.

.. note:: 
   
   For more information about subdocument matching, see the
   `Query on Embedded/Nested Documents <https://www.mongodb.com/docs/manual/tutorial/query-embedded-documents/>`__
   guide in the MongoDB Server documentation.

> What Does "*CursorNotFound* cursor id not valid at server" Mean?

Author: moved to troubleshooting

> No. We've decided not to implement this feature, for the following reasons:

Author: thought i could keep it more informal on the FAQ page, and also wanted to indicate this was not a change that would likely be coming in the future. your version is good too though

> 1. It would pollute the attribute namespace for documents and could

Reviewer: ```suggestion
1. It pollutes the attribute namespace for documents and could
```

Does MongoDB Support Attribute-style Access for Documents?
----------------------------------------------------------

We've decided not to implement anything like this. The relevant `Jira case
<http://jira.mongodb.org/browse/PYTHON-35>`__ has some information
about the decision, but here is a brief summary:
No. We've decided not to implement this feature, for the following reasons:

1. This will pollute the attribute namespace for documents and could
1. It would pollute the attribute namespace for documents and could
   lead to subtle bugs or confusing errors when using a key with the
   same name as a dictionary method.

2. The only reason we even use SON objects instead of regular
   dictionaries is to maintain key ordering, since the server
   requires this for certain operations. We're hesitant to
   needlessly complicate SON, since at some point
   we might want to revert back to using dictionaries alone
   without breaking backwards compatibility.

3. It's easy (and Pythonic) for new users to deal with documents,
   since they behave just like dictionaries. If we start changing
   their behavior, it adds a barrier to entry for new users (another
   class to learn).

> What is the Correct Way to Handle Time Zones with {+driver-short+}?

Author: moved to date-time page

----------------------------------------------------------

See :ref:`pymongo-dates-times` for examples of how to handle
``~datetime.datetime`` objects correctly.

How Can I Save a ``datetime.date`` Instance?
-----------------------------------------------

{+driver-short+} doesn't support saving ``datetime.date`` instances, since
there is no BSON type for dates without times. Rather than having the
driver enforce a convention for converting ``datetime.date``
instances to ``datetime.datetime`` instances for you, you should
perform any conversion in your client code.

.. _web-application-querying-by-objectid:

> Why Do I Get No Result When I Query for a Document by ObjectId in My Web Application?

Author: moved to troubleshooting

It's common in web applications to encode documents' ObjectIds in URLs, as shown
in the following code example:

.. code-block:: python

   "/posts/50b3bda58a02fb9a84d8991e"

Your web framework will pass the ObjectId portion of the URL to your request
handler as a string, so it must be converted to a ``~bson.objectid.ObjectId`` instance
before it's passed to the ``Collection.find_one`` method. It's a
common mistake to forget to do this conversion. Here's how to do it correctly
in `Flask <http://flask.pocoo.org/>`__ (other web frameworks are similar):

.. code-block:: python

   from pymongo import MongoClient
   from bson.objectid import ObjectId

   from flask import Flask, render_template

   client = MongoClient()
   app = Flask(__name__)

   @app.route("/posts/<_id>")
   def show_post(_id):
      # NOTE!: converting _id from string to ObjectId before passing to find_one
      post = client.db.posts.find_one({'_id': ObjectId(_id)})
      return render_template('post.html', post=post)

   if __name__ == "__main__":
       app.run()

.. note::
   
   For more information, see the :ref:`querying-by-objectid` section in the Find guide.

> How Can I Use {+driver-short+} from Django?

Author: moved to tools

> Yes. See :ref:`pymongo-mod_wsgi` in the Tools guide.

Author: good catch

For more information, see `PYTHON-1495 <https://jira.mongodb.org/browse/PYTHON-1495>`__ in
the Python Driver Jira project.

How Can I Use Something Like Python's ``json`` Module to Encode My Documents to JSON?
-------------------------------------------------------------------------------------

``~bson.json_util`` is {+driver-short+}'s built-in, flexible tool for using
Python's ``json`` module with BSON documents and `MongoDB Extended JSON
<https://mongodb.com/docs/manual/reference/mongodb-extended-json/>`__. The
``json`` module won't work out of the box with all documents from {+driver-short+},
as {+driver-short+} supports some special types (like ``~bson.objectid.ObjectId``
and ``~bson.dbref.DBRef``) that are not supported in JSON.

`python-bsonjs <https://pypi.python.org/pypi/python-bsonjs>`__ is a fast
BSON-to-MongoDB-Extended-JSON converter built on top of
`libbson <https://github.com/mongodb/libbson>`__. ``python-bsonjs`` doesn't
depend on {+driver-short+} and can offer a performance improvement over
``~bson.json_util``. ``python-bsonjs`` works best with {+driver-short+} when using
``~bson.raw_bson.RawBSONDocument``.

> Why Do I Get OverflowError When Decoding Dates Stored by Another Language's Driver?

Author: troubleshooting

> with subtype 0. the byte string is decoded to ``Binary``.

Reviewer: ```suggestion
with subtype 0. In Python 2, the byte string is decoded to ``Binary``.
```

   >>> pickle.dumps(oid)
   'ccopy_reg\n_reconstructor\np0\n(cbson.objectid\...'

   # Python 3.7
   >>> import pickle
   >>> pickle.loads(b'ccopy_reg\n_reconstructor\np0\n(cbson.objectid\...', encoding='latin-1')
   ObjectId('4f919ba2fba5225b84000000')

If you pickled an ``ObjectID`` in Python 2, and want to unpickle it in Python 3,

> you must pass the ``protocol <= 2`` option to the ``pickle.dumps()`` method.

Reviewer: This line says to pass `protocol <= 2` to the method, but in the highlighted code block it passes `protocol=2`. I'm not sure which one is correct, or if I'm misunderstanding something with it, but we should probably clean that up
